botocore/parsers.py

```diff
@@ -658,16 +649,29 @@ def _do_parse(self, response, shape):
         return final_parsed
 
     def _do_error_parse(self, response, shape):
-        error = {
-            'Error': {
-                'Code': response['headers'].get(':error-code', ''),
-                'Message': response['headers'].get(':error-message', ''),
+        exception_type = response['headers'].get(':exception-type')
+        exception_shape = shape.members.get(exception_type)
+        if exception_shape is not None:
+            body = {}
+            members = exception_shape.members
+            self._parse_payload(response, exception_shape, members, body)
+            error = {
+                'Error': {
+                    'Code': exception_type,
+                    'Message': body.get('Message', body.get('message', ''))
+                }
+            }
+        else:
+            error = {
+                'Error': {
+                    'Code': response['headers'].get(':error-code', ''),
+                    'Message': response['headers'].get(':error-message', ''),
+                }
             }
-        }
         return error
 
     def _parse_payload(self, response, shape, member_shapes, final_parsed):
-        if shape.serialization.get('event'):
+        if shape.serialization.get('event') or shape.metadata.get('exception'):
```

Comment:
It doesn't seem like the `event` and `exception` code path should be consolidated:

* The `exception` shapes would never have an `eventpayload` so that code path will never be reached.
* We don't actually need to parse the error body according to its shape because `_do_error_parse` only checks for a `message` key.  Anything else it parses it just throws away.

Seems like it's simpler to just call `self._initial_body_parse` in `_do_error_parse`, and if you really wanted to be safe, a subsequent call to`self._parse_shape`.

Reply:
To be safe I'll still parse the body according to the shape.